src/ls/definition.cr

```diff
@@ -0,0 +1,166 @@
+module Mint
+  module LS
+    class StackReader
+      def initialize(@stack : Array(Ast::Node))
+        @index = 0
+      end
+
+      def find_next(klass : N.class) : N | Nil forall N
+        node = @stack[@index]?.try &.as?(N)
+        @index += 1
+        node
+      end
+
+      def find_anywhere(klass : N.class) : N | Nil forall N
+        @stack[@index..].each do
+          node = find_next(klass)
+          return node if node
+        end
+      end
+    end
+
+    # This is the class that handles the "textDocument/definition" request.
+    class Definition < LSP::RequestMessage
+      property params : LSP::TextDocumentPositionParams
+
+      def execute(server) : LSP::LocationLink | LSP::Location | Nil
+        uri =
+          URI.parse(params.text_document.uri)
+
+        workspace =
+          Workspace[uri.path.to_s]
+
+        unless workspace.error
+          stack =
+            server.nodes_at_cursor(params)
+
+          html_style(server, workspace, stack) ||
+            html_attribute(server, workspace, stack) ||
+            html_component(server, workspace, stack)
+        end
+      end
+
+      def with_stack(stack : Array(Ast::Node), &)
+        yield StackReader.new(stack)
+      end
+
+      def selection(location : Ast::Node::Location) : LSP::Range
+        LSP::Range.new(
+          start: LSP::Position.new(
+            line: location.start[0] - 1,
+            character: location.start[1]
+          ),
+          end: LSP::Position.new(
+            line: location.end[0] - 1,
+            character: location.end[1]
+          )
+        )
+      end
+
+      # Returns the range for the name part for a node
+      def selection(node : Ast::Node) : LSP::Range
+        selection(node.location)
+      end
+
+      def selection(node : Ast::Component) : LSP::Range
+        # Select only the name part of the component
+        #   global component MintComponent {
+        #                    ^^^^^^^^^^^^^
+
+        start_line, start_column = node.location.start
+
+        node.comment.try do |comment|
+          start_line, start_column = comment.location.end
+        end
+
+        # TODO: Change parser so component name is a node?
```

Comment:
You can change the parser if you can :)

Reply:
Would certainly simplify this method :smiley: 

Looking at it, the best way might be to have the `type_id` methods return a new node (`TypeId`?) rather than a `String`?

Reply:
Oh I forgot that it was a `String`, then yes `TypeId` is what we want. 

We can do it in a separate PR as well, it's up to you 🙂

Reply:
Definitely, this PR is big enough as it is 😂